### ISSUE TITLE
Useless calls from the embedder

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#25305]
Follow up of #25263.

The number of calls from the embedder to Servo can be absurdly high. 300K+ per seconds.

The way we let the embedder know about the running animation state and run animations is spaghetti-esque. A mix of #21855 and #20379 issues.

This is how we would expect animations to work:

1. animation happens Layout
2. let the compositor know
3. let the embedder know
4. embedder runs full speed and block on swapbuffer.

But:

Step 2, letting the compositor know, triggers the compositor proxy to wake up the embedder loop which will run a `compositor.perform_update()` call which triggers compositing.
Now goes step 3, animating is set to animating.
Step 4, compositor won't swap the buffer because nothing needs to be done, compositing has already happened in step 2.

Adding to that the fact that AnimationState is sent at each tick, and we don't end the animation properly, and there is nothing to animate, the embedder calls perform_update full speed, and block on nothing, ending up with 300K+ calls per seconds.

**Comment by paulrouget** [comment]
After cleaning up a bit the way animations state are sent, we see that animation are actually stopped and then restarted.

**Comment by paulrouget** [comment]
Step one: #25306

**Comment by paulrouget** [comment]
I'm not going to invest too much time in understanding the CSS animation issue. https://github.com/servo/servo/issues/20379#issuecomment-566137186 gives some clues and https://github.com/servo/servo/pull/25306 helps a lot.

**Comment by paulrouget** [comment]
(closed by mistake)

For WebGL, I'm not exactly sure what would be the best approach.

A simple example of a problematic web page:
```html
  <script>
    function foo() {
      requestAnimationFrame(foo);
    }
    foo();
  </script>
```

Because there is not call to swap_buffer, this will loop full speed.

I think we should track if call to `servo::handle_events` ends up calling `window::present`, and if not, sleep the GL thread for ~1ms (or even longer).

At the end, I'm not sure it's a good idea to let the embedder loop on animation. Maybe servo could do the scheduling itself by waking up the compositor when needed.

**Comment by paulrouget** [comment]
Note about blocking for 1ms: we will need to make sure it doesn't slow down WebXR.

**Comment by pshaughn** [comment]
https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering the concept of "rendering opportunity" is vague in step 2 here, but in practice it seems to be consistently interpreted as  meaning rate-limit requestAnimationFrame to 60 frames per second or slower.

**Comment by gterzian** [comment]
Sounds like another symptom of the general problem of https://github.com/servo/servo/issues/23899

**Comment by gterzian** [comment]
@paulrouget Can you not replace "animation mode" where the embedder is running full-speed with something like "maybe run animations" each time script requests another animation frame?

**Comment by paulrouget** [comment]
@gterzian Yes. I'd like to split the perform_update logic in 2. Having something specific for animation will help. Having so many thing under perform_update make the logic indeed too complex.

For now I'll add the speed limit to the compositor as it's necessary not matter what.

**Comment by asajeffrey** [comment]
I'm with @gterzian here, really we should redesign the embedder API, especially since surfman allows us to move servo off the main thread. In the mean time, I'm still confused why `swap_buffers` isn't being called, is this in the case where the compositor runs but there's been no change, so `swap_buffers` doesn't happen?

**Comment by paulrouget** [comment]
> is this in the case where the compositor runs but there's been no change, so swap_buffers doesn't happen?

Yes.

> especially since surfman allows us to move servo off the main thread

Would that work on Mac for example?
Could we call swap_buffers in a different thread?

> really we should redesign the embedder API

Absolutely. The current PR is really a temporary fix for the HoloLens.

If we are fine with the extra calls for now, I will prioritize the re-factoring work of the embedding API.

**Comment by asajeffrey** [comment]
Yes, this should work for any architecture that can support surfman. The downside is an extra blit :(

**Comment by paulrouget** [comment]
@gterzian :
> Can you replace "animation mode" where the embedder is running full-speed with something like "maybe run animations" each time script requests another animation frame, so that instead of running full-speed it runs at the speed of script requesting frames, and only if the compos
itor deems it necessary

I'm not sure to understand how that would work. When to call the callback?

It's important to understand that swap_buffer blocks long enough to force the cadence at 60 FPS. And swap_buffers is only called if there's something to draw.

@asajeffrey:
> I'm with @gterzian here, really we should redesign the embedder API

There's the design of the API itself (the cosmetic part), which is not in a great shape, and there's the way the threads work. To solve this specific issue, I don't think redesigning the API itself will help.

-----

1) what would be the alternative to pausing the compositor thread
2) how the embedder API needs to change (excluding cosmetic changes).
3) what needs to be done right now, what can wait

For 1), I don't think it's possible to not pause somewhere.
For 2), the only change I can imagine is to split embedder and compositor in 2 different threads. And we will end up pausing the compositor thread anyway,
So, for 3) can we land the proposed PR, as it immediatly improves the HL situation, and we will end-up pausing anyway. And then, I start moving the compositor to its own thread. From that thread we will have a system to make sync calls to swap_buffer/present.

**Comment by gterzian** [comment]
What I mean is the following:

We now have this in the API: https://github.com/servo/servo/blob/master/ports/libsimpleservo/api/src/lib.rs#L131

Which will result in calling `peform_updates` in a loop. The doc comment assumes that this will result in 60FPS because of blocking on swapping buffers, but you wrote earlier that his doesn't actually happen as expected. 

So, since if we are animating something, that means script is regularly requesting new animation frames, right? For example for the "normal" window animation frames, the document sends a `ScriptMsg::ChangeRunningAnimationsState(AnimationState::AnimationCallbacksPresent)`, which is then forwarded by the constellation to the compositor. 

Forwarding that message from the constellation to the compositor wakes up the main event-loop. 

Is that wake-up perhaps occurring frequently enough to get 60FPS, without looping at full speed? In other words, does the requesting of animation frames not provide a natural way for the embedder event-loop to run fast enough during animations?

**Comment by paulrouget** [comment]
@gterzian ok - I understand what you’re talking about. This is #21855. We could indeed do without looping. But, that doesn’t matter that much beyond the “cosmetic part” of it.

We will still have to pause, in the compositor thread, if a swap_buffer doesn’t occur.

**Comment by paulrouget** [comment]
3 improvements on top of #25312:

1. only pause if swap_buffer was not called, as opposed to checking if the frame duration was too short
2. the period of time we wait for when swap_buffer didn’t happen should not be arbitrary
3. the compositor should be in its own thread

I’ll go ahead and do that in that order.

**Comment by paulrouget** [comment]
Related: https://github.com/servo/webxr/issues/113#issuecomment-574399195

**Comment by paulrouget** [comment]
Looking at Gecko's code shows that Gecko either blocks for a bit or relies on platform's specific implementation to get notified for the next sync:

* [wayland](https://searchfox.org/mozilla-central/rev/0439db3a39faddb90197a87cc184c57dcbf0a770/widget/gtk/WaylandVsyncSource.cpp#56)
* [windows](https://searchfox.org/mozilla-central/rev/0439db3a39faddb90197a87cc184c57dcbf0a770/gfx/thebes/gfxWindowsPlatform.cpp#1844)
* [mac](https://searchfox.org/mozilla-central/rev/0439db3a39faddb90197a87cc184c57dcbf0a770/gfx/thebes/gfxPlatformMac.cpp#442)
* [gtk](https://searchfox.org/mozilla-central/rev/0439db3a39faddb90197a87cc184c57dcbf0a770/gfx/thebes/gfxPlatformGtk.cpp#629)
* [android](https://searchfox.org/mozilla-central/rev/0439db3a39faddb90197a87cc184c57dcbf0a770/gfx/thebes/gfxAndroidPlatform.cpp#283)

For each platform, a specific vsync thread runs and sends a signal on vsync.

**Comment by paulrouget** [comment]
Note: the current implementation has a "work around" for no-op rAF: https://github.com/servo/servo/blob/5597ccf57ddc2b77fcb4a8071f575b0dc9389a12/components/script/dom/document.rs#L1641

We really need to remove that.

**Comment by paulrouget** [comment]
So initially, #14442 introduced a workaround to avoid an avalanche of rAF, by detecting calls to rAF that would not create a DOM update. So even if the embedder itself would go full speed, the rAF would be limited.

I believe this throttling should happen at the compositor level, as it will also limit the number of calls to the embedder.

I've tried to:
1. remove the fake rAF
2. detect when a composite call end up not compositing
3. if no compositing, pause long enough to make the frame last 16ms

This obviously improves the situation a lot, but for some reason, we reach 75FPS (consistently), not 60FPS. And even using a proper animation (updating the layout on each rAF), we still end-up artificially pausing.

Still need more investigation on that front.

Moving the compositor to its own thread and having a scheduler [like in gecko](https://github.com/servo/servo/issues/25305#issuecomment-587334896) will be the logical next step.

**Comment by paulrouget** [comment]
So Glutin 0.21 runs full speed even with vsync enabled. The latest version of glutin appears to have a different callback system that might fix this issue.

This doesn't help…

**Comment by paulrouget** [comment]
As I'm working on this, I'm finding some more issues.

On Windows, with the UWP app, the cadencing is good at the embedder level. But the script_thread can't follow. Often [skipping](https://github.com/servo/servo/blob/12693b51f510761d5dd58a9103596a192edb803c/components/script/script_thread.rs#L1490) messages from the compositor.

Not sure why yet.

**Comment by paulrouget** [comment]
More investigation:

- about to redraw
- composite (swap_buffer)
- send TickAnim
- looping back
- about to redraw
- the script didn't have the time propagate the new layout changes to WR
- composite missed
- send TickAnim
- barely no time happened since the first TickAnim, script is still processing events, skipping the rAF callback
- embedder figure out missed composite
- sleep
- script finally got enough time to propagate the changes to WR
- looping back

Missing half of the TickAnim.

**Comment by paulrouget** [comment]
The main issue we have here is that we rely on `swap_buffer` to run at 60FPS. Which sounds like the right thing to do, but we do not always have something to paint, so `swap_buffer` is not always called, so nothing is blocking, making the loop going too fast.

We still need to call the rAF callback at 60FPS, and also tick the layout thread.

I would like to propose this change:

- do not loop if the compositor is animating (remove that "AnimationState" mode)
- do not block on swapbuffer.
- have a dedicated thread that ticks on vblank
- that thread would notify script and layout (via the compositor maybe, to make sure we notify the visible pipelines)
- if something needs to be painted, the compositor will be notified via the regular route
- remove the spurious animation frame check

This will also work around the Glutin loop going too fast, and make it a lot easier to split the embedder/compositor code in 2 threads.

The difficulty will be to build a cross platform vsync notifier. As a temporary solution, we can start with a simple timer, and platform by platform, properly implement the scheduler (see what gecko does [here](https://github.com/servo/servo/issues/25305#issuecomment-587334896)).

@pcwalton I'd like to have your opinion on this.

**Comment by gterzian** [comment]
> As a temporary solution, we can start with a simple timer

Maybe you can start with adding a channel in one of the existing selects, see for example how timers are implemented in the constellation: https://github.com/servo/servo/blob/f5ff38b87559cd25db44fa7bc6f2a9c4cb2264e0/components/constellation/constellation.rs#L1378

How about first moving the compositor to it's own thread, and then have the timer be a channel included in the select?  Then the other channels could represent the "other route", which I assume means the raf loops? 

Also, and this sort of depends on whether my mental model of what you are doing is right, could you not actually use the timer in the following way:

```rust 

// This runs in a dedicated compositor thread.

 let rafs = VecDeque::new();

loop {
    // Replace this with a per-platform source of vsync events.
    let next_render = after(timeout);

    select! {
        recv(self.window_raf_receiver) -> req => rafs.push_back(Raf::FromWindow(req)),
        recv(self.xr_raf_receiver) -> req => rafs.push_back(Raf::FromXr(req)),
        recv(self.worker_raf_receiver) -> req => afs.push_back(Raf::FromWorker(req)),
        recv(next_render) -> _  => {
               self.composite();
               for raf in rafs.drain(0..) {
                    self.tick_animations(raf);
               }
        }
}
}
```

**Comment by mrobinson** [comment]
Things are a bit better now as the compositor will not tick animations as fast as possible. Instead it uses a timer, but that should eventually be driven by vsync.

**Comment by mrobinson** [comment]
I think we can consider this complete after #39072.
